File: bongf/java/src/level2/PhoneNumberList.java
package level2;

import java.util.*;

public class PhoneNumberList {
    public static void main(String[] args) {
        System.out.println(solution(new String[]{"97674223", "1195524421", "119"}));
        System.out.println(solution(new String[]{"1195524421", "97674223", "1"}));
        System.out.println(solution(new String[]{"010", "011", "0112", "112"}));
        System.out.println(solution2(new String[]{"119", "97674223", "1195524421"}));
    }

    public static boolean solution(String[] phone_book) {
        Map<Character, List<String>> dic = new HashMap<>();
        dic.put('0', new ArrayList<>());
        dic.put('1', new ArrayList<>());
        dic.put('2', new ArrayList<>());
        dic.put('3', new ArrayList<>());
        dic.put('4', new ArrayList<>());
        dic.put('5', new ArrayList<>());
        dic.put('6', new ArrayList<>());
        dic.put('7', new ArrayList<>());
        dic.put('8', new ArrayList<>());
        dic.put('9', new ArrayList<>());

        for( String num : phone_book) {
            dic.get(num.charAt(0)).add(num);
        }

        for (Character character : dic.keySet()) {
            List<String> ns = dic.get(character);
            if(ns.size() >= 2) {
                Collections.sort(ns);
                for (int i = 0; i < ns.size() - 1; i++) {
                    if(ns.get(i+1).startsWith(ns.get(i))) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public static boolean solution2(String[] phone_book) {
        Arrays.sort(phone_book);
        if(phone_book.length >= 2) {
            for (int i = 0; i < phone_book.length-1; i++) {
                if (phone_book[i+1].startsWith(phone_book[i])) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: programmers/python/level2/phone_number_list.py
def solution(phone_book):
    dic = {}
    dic = {'0':[], '1':[],'2':[],'3':[],'4':[],'5':[],'6':[],'7':[],'8':[],'9':[]}

    for num in phone_book :
        dic[num[0]].append(num)
    
    for i in range(10):
        if len(dic[str(i)]) >= 2 :
            dic[str(i)].sort()
            for j in range(len(dic[str(i)])-1) :
                if dic[str(i)][j+1].startswith(dic[str(i)][j]) :
                    return False
    return True

def solution(phoneBook):
    phoneBook = sorted(phoneBook)

    if(len(phoneBook) >= 2) :
        for p1, p2 in zip(phoneBook, phoneBook[1:]):
            if p2.startswith(p1):
                return False
    return True

print(solution(["97674223", "1195524421", "119"]))
print(solution(["1195524421", "97674223", "1"]))
print(solution(["010", "011", "0112", "112"]))




